### PR TITLE
Add ccache-toolchain

### PR DESCRIPTION
**Pull request by isuruf**
cc @gqmelo

### CODE REVIEW

File: recipes/ccache-toolchain/activate.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+export _OLD_PATH=${PATH}
```
Comment (gqmelo) [inline recipes/ccache-toolchain/activate.sh]: There is a problem with this approach. The order which activate/deactivate scripts is undefined.
So if you have other package that changes the PATH, ```_OLD_PATH``` can contain the wrong value.

For example:

```foo``` has an activate script that does:

```
export PATH="$FOO_DIR:$PATH"
```
and bar:


```
export _OLD_PATH="$PATH"
export PATH="$BAR_DIR:$PATH"
```

If foo's activate is executed first ```_OLD_PATH``` will contain ```$FOO_DIR```. Then if foo's deactivate is executed first, it will remove ```$FOO_DIR``` from ```PATH```, but on bar's deactivate it will revert ```PATH``` to ```_OLD_PATH```, which will still contain ```$FOO_DIR```

So I suggest doing this approach: https://github.com/gqmelo/gcc-recipes/blob/master/libgcc/deactivate-gcc-libs.sh
Also it would be good to copy the relevant tests as this is tricky to get it right: https://github.com/gqmelo/gcc-recipes/blob/master/libgcc/run_test.py

File: recipes/ccache-toolchain/activate.sh [hunk 2]
```diff
@@ -2,3 +2,11 @@
 
 export _OLD_PATH=${PATH}
 export PATH="${PREFIX}/bin/conda_forge_ccache:${PATH}"
+
+if [ -f "/feedstock_root/build_artefacts" ]; then
+    export CCACHE_DIR=/feedstock_root/build_artefacts/.ccache
+fi
+
+# Set max cache size so we don't carry old objects for too long
+ccache -M 400M
```
Comment (isuruf) [inline recipes/ccache-toolchain/activate.sh]: @gqmelo, for sagelib I need around 350M on osx. That's why I had it like this. What would happen if this is set to 200M and in sagelib package it is set to 400M? Will setting to 200M here, immediately downsize to 200M?

File: recipes/ccache-toolchain/activate.sh [hunk 3]
```diff
@@ -2,3 +2,11 @@
 
 export _OLD_PATH=${PATH}
 export PATH="${PREFIX}/bin/conda_forge_ccache:${PATH}"
+
+if [ -f "/feedstock_root/build_artefacts" ]; then
```
Comment (isuruf) [inline recipes/ccache-toolchain/activate.sh]: This is done to make sure the recipes work correctly locally.

File: recipes/ccache-toolchain/deactivate.sh [hunk 4]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+export PATH="${_OLD_PATH}"
+unset CCACHE_DIR
```
Comment (gqmelo) [inline recipes/ccache-toolchain/deactivate.sh]: I think it's useful to add a ```ccache -s``` here. Specially on the beginning it will be very helpful to see the statistics to make sure everything is working.

File: recipes/ccache-toolchain/activate.sh (hunk 3 above)
Comment (gqmelo) [inline recipes/ccache-toolchain/activate.sh]: Great! This solves the problem

File: recipes/ccache-toolchain/activate.sh (hunk 2 above)
Comment (gqmelo) [inline recipes/ccache-toolchain/activate.sh]: I think ccache will clear the exceeding size later when you try to compile something, but I'm not sure. We need to check that

File: recipes/ccache-toolchain/activate.sh (hunk 1 above)
Comment (isuruf) [inline recipes/ccache-toolchain/activate.sh]: I thought the activate and deactivate happened in a way such that the last activate is deactivated first. Let me check that

Comment (gqmelo) [inline recipes/ccache-toolchain/activate.sh]: Yeah, me too. But I was bitten by that before. Conda just list the files and the order is undefined (at least on ext4 filesystems)

File: recipes/ccache-toolchain/activate.sh (hunk 2 above)
Comment (isuruf) [inline recipes/ccache-toolchain/activate.sh]: Just checked and yes it will clear (not all at first) when you try to compile something.

File: recipes/ccache-toolchain/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    - ccache      # [unix]
+    - toolchain 2.*
+
+test:
+  commands:
+    # Verify the scripts are in-place.
+    {% for state in ["activate", "deactivate"] %}
+    - test -f "${PREFIX}/etc/conda/{{ state }}.d/ccache_toolchain_{{ state }}.sh"  # [unix]
+    {% endfor %}
+    {% for cc in ["gcc", "g++", "clang", "clang++", "gfortran"] %}
```
Comment (gqmelo) [inline recipes/ccache-toolchain/meta.yaml]: It would be nice to add some test executing the symlinks. I think just a running them with ```--help``` would be enough

Comment (isuruf) [inline recipes/ccache-toolchain/meta.yaml]: Done in 505d5cb

File: recipes/ccache-toolchain/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+requirements:
+  build:
+    - ccache      # [unix]
+  run:
+    - ccache      # [unix]
+    - toolchain 2.*
+
+test:
+  requires:
+    - gcc         # [unix]
```
Comment (gqmelo) [inline recipes/ccache-toolchain/meta.yaml]: Why did you need the gcc package?

Comment (gqmelo) [inline recipes/ccache-toolchain/meta.yaml]: Ah, ok, you are running gcc and g++ on MacOS

**Comment by isuruf** [comment]
Thanks @gqmelo for the review

**Comment by bsschwar** [comment]
I'm a bit confused how this works. Is there a document that would tell me how to use this? I did `conda install -c conda-forge ccache'
'conda install -c conda-forge ccache-toolchain'
 'which gcc` and it still shows me /usr/bin/gcc instead of a link to ccache.

**Comment by isuruf** [comment]
This is not complete yet. (Needs a patch to conda-smithy) We'll add some docs when this is done.
